File: libsignal-service/src/main/java/org/whispersystems/signalservice/api/storage/SignalStorageManifest.java
package org.whispersystems.signalservice.api.storage;

import org.signal.core.util.ProtoUtil;
import org.whispersystems.signalservice.internal.storage.protos.ManifestRecord;
import org.whispersystems.signalservice.internal.storage.protos.StorageManifest;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

import okio.ByteString;

public class SignalStorageManifest {
  public static final SignalStorageManifest EMPTY = new SignalStorageManifest(0, 1, Collections.emptyList());

  private final long                          version;
  private final int                           sourceDeviceId;
  private final List<StorageId>               storageIds;
  private final Map<Integer, List<StorageId>> storageIdsByType;

  public SignalStorageManifest(long version, int sourceDeviceId, List<StorageId> storageIds) {
    this.version          = version;
    this.sourceDeviceId   = sourceDeviceId;
    this.storageIds       = storageIds;
    this.storageIdsByType = new HashMap<>();

    for (StorageId id : storageIds) {
      List<StorageId> list = storageIdsByType.get(id.getType());
      if (list == null) {
        list = new ArrayList<>();
      }
      list.add(id);
      storageIdsByType.put(id.getType(), list);
    }
  }

  public static SignalStorageManifest deserialize(byte[] serialized) {
    try {
      StorageManifest manifest       = StorageManifest.ADAPTER.decode(serialized);
      ManifestRecord  manifestRecord = ManifestRecord.ADAPTER.decode(manifest.value_);
      List<StorageId> ids            = new ArrayList<>(manifestRecord.identifiers.size());

      for (ManifestRecord.Identifier id : manifestRecord.identifiers) {
        ids.add(StorageId.forType(id.raw.toByteArray(), id.type.getValue()));
      }

      return new SignalStorageManifest(manifest.version, manifestRecord.sourceDevice, ids);
    } catch (IOException e) {
      throw new AssertionError(e);
    }
  }

  public long getVersion() {
    return version;
  }

  public int getSourceDeviceId() {
    return sourceDeviceId;
  }

  public String getVersionString() {
    return version + "." + sourceDeviceId;
  }

  public List<StorageId> getStorageIds() {
    return storageIds;
  }

  public Optional<StorageId> getAccountStorageId() {
    List<StorageId> list = storageIdsByType.get(ManifestRecord.Identifier.Type.ACCOUNT.getValue());

    if (list != null && list.size() > 0) {
      return Optional.of(list.get(0));
    } else {
      return Optional.empty();
    }
  }

  public Map<Integer, List<StorageId>> getStorageIdsByType() {
    return storageIdsByType;
  }

  public byte[] serialize() {
    List<ManifestRecord.Identifier> ids = new ArrayList<>(storageIds.size());

    for (StorageId id : storageIds) {
      ManifestRecord.Identifier.Type type = ManifestRecord.Identifier.Type.Companion.fromValue(id.getType());
      if (type != null) {
        ids.add(new ManifestRecord.Identifier.Builder()
                                             .type(type)
                                             .raw(ByteString.of(id.getRaw()))
                                             .build());
      } else {
        ByteString unknownEnum = ProtoUtil.writeUnknownEnumValue(StorageRecordProtoUtil.STORAGE_ID_TYPE_TAG, id.getType());
        ids.add(new ManifestRecord.Identifier(ByteString.of(id.getRaw()), ManifestRecord.Identifier.Type.UNKNOWN, unknownEnum));
      }
    }

    ManifestRecord manifestRecord = new ManifestRecord.Builder()
                                                      .identifiers(ids)
                                                      .sourceDevice(sourceDeviceId)
                                                      .build();

    return new StorageManifest.Builder()
                              .version(version)
                              .value_(manifestRecord.encodeByteString())
                              .build()
                              .encode();
  }
}

File: libsignal-service/src/main/java/org/whispersystems/signalservice/api/storage/SignalStorageManifest.kt
package org.whispersystems.signalservice.api.storage

import okio.ByteString
import okio.ByteString.Companion.EMPTY
import okio.ByteString.Companion.toByteString
import org.signal.core.util.ByteSize
import org.signal.core.util.bytes
import org.signal.core.util.isNotEmpty
import org.signal.core.util.toOptional
import org.whispersystems.signalservice.internal.storage.protos.ManifestRecord
import org.whispersystems.signalservice.internal.storage.protos.StorageManifest
import java.util.Optional

data class SignalStorageManifest(
  @JvmField val version: Long,
  val sourceDeviceId: Int,
  val recordIkm: RecordIkm?,
  @JvmField val storageIds: List<StorageId>,
  val protoByteSize: ByteSize
) {

  constructor(version: Long, sourceDeviceId: Int, recordIkm: RecordIkm?, storageIds: List<StorageId>) : this(
    version = version,
    sourceDeviceId = sourceDeviceId,
    recordIkm = recordIkm,
    storageIds = storageIds,
    protoByteSize = toProto(version, storageIds, sourceDeviceId, recordIkm).encode().size.bytes
  )

  companion object {
    val EMPTY: SignalStorageManifest = SignalStorageManifest(0, 1, null, emptyList())

    fun deserialize(serialized: ByteArray): SignalStorageManifest {
      val manifest = StorageManifest.ADAPTER.decode(serialized)
      val manifestRecord = ManifestRecord.ADAPTER.decode(manifest.value_)
      val ids: List<StorageId> = manifestRecord.identifiers.map { id ->
        StorageId.forType(id.raw.toByteArray(), id.typeValue)
      }

      return SignalStorageManifest(
        version = manifest.version,
        sourceDeviceId = manifestRecord.sourceDevice,
        recordIkm = manifestRecord.recordIkm.takeIf { it.isNotEmpty() }?.toByteArray()?.let { RecordIkm(it) },
        storageIds = ids,
        protoByteSize = serialized.size.bytes
      )
    }

    private fun toProto(version: Long, storageIds: List<StorageId>, sourceDeviceId: Int, recordIkm: RecordIkm?): StorageManifest {
      val ids: List<ManifestRecord.Identifier> = storageIds.map { id ->
        ManifestRecord.Identifier.fromPossiblyUnknownType(id.type, id.raw)
      }

      val manifestRecord = ManifestRecord(
        identifiers = ids,
        sourceDevice = sourceDeviceId,
        recordIkm = recordIkm?.value?.toByteString() ?: ByteString.EMPTY
      )

      return StorageManifest(
        version = version,
        value_ = manifestRecord.encodeByteString()
      )
    }
  }

  val storageIdsByType: Map<Int, List<StorageId>> = storageIds.groupBy { it.type }

  val versionString: String
    get() = "$version.$sourceDeviceId"

  val accountStorageId: Optional<StorageId>
    get() = storageIdsByType[ManifestRecord.Identifier.Type.ACCOUNT.value]?.takeIf { it.isNotEmpty() }?.get(0).toOptional()

  fun serialize(): ByteArray {
    val ids: List<ManifestRecord.Identifier> = storageIds.map { id ->
      ManifestRecord.Identifier.fromPossiblyUnknownType(id.type, id.raw)
    }

    val manifestRecord = ManifestRecord(
      identifiers = ids,
      sourceDevice = sourceDeviceId,
      recordIkm = recordIkm?.value?.toByteString() ?: ByteString.EMPTY
    )

    return StorageManifest(
      version = version,
      value_ = manifestRecord.encodeByteString()
    ).encode()
  }
}
